torch/csrc/dynamo/guards.cpp

```diff
@@ -574,6 +574,480 @@ static struct PyModuleDef _module = {
+  // purpose.
+  bool check(py::handle value) {
+    return check_nopybind(value.ptr());
+  }
+
+  GuardDebugInfo debug_check(py::handle value) {
+    return debug_check_nopybind(value.ptr());
+  }
+
+  GuardDebugInfo debug_check_nopybind(PyObject* value) { // borrowed ref
+    bool result = check_nopybind(value);
+    std::string failure_reason = "";
+    if (!result) {
+      failure_reason = get_failure_reason(value);
+    }
+    return GuardDebugInfo(result, failure_reason, 0);
+  }
+
+  // This is on the hot path and avoids any refcounting code from pybind. This
+  // is not exposed to Python and can only be called from C++.
+  virtual bool check_nopybind(PyObject* value) = 0;
+  virtual std::string get_failure_reason(PyObject* value) = 0;
+  virtual ~LeafGuard() = default;
+};
+
+/**
+ * Represents a leaf guard that accepts the python guard check function. We
+ * would like to have most of the guards in C++ (to avoid a Python function
+ * call).  But, it will take some time to reach that goal. Also, there might be
+ * cases where its too tedious to write an equivalent C++ guard.
+ *
+ * PythonLambdaGuard allows us to gradually move to C++. We can start from all
+ * guards of type PythonLambaGuard and incrementally move expensive guards to
+ * C++.
+ */
+class PythonLambdaGuard : public LeafGuard {
+ public:
+  PythonLambdaGuard(py::object guard_check_fn, py::object print_failure_fn) {
+    if (py::isinstance<py::function>(guard_check_fn) &&
+        py::isinstance<py::function>(print_failure_fn)) {
+      _guard_check_fn = py::cast<py::function>(guard_check_fn);
+      _guard_check_fn_pyobj = _guard_check_fn.ptr();
+      _print_failure_fn = py::cast<py::function>(print_failure_fn);
+    } else {
+      throw py::type_error("PythonLambdaGuard expects callables");
+    }
+  }
+
+  // Runs the lambda function with the current f_locals value.
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    return PyObject_IsTrue(PyObject_CallOneArg(_guard_check_fn_pyobj, value));
+  }
+
+  std::string get_failure_reason(PyObject* value) override {
+    return py::cast<std::string>(_print_failure_fn(py::handle(value)));
+  }
+
+ private:
+  // The user provided lambda function for check_fn.
+  py::function _guard_check_fn;
+  PyObject* _guard_check_fn_pyobj;
+  // The user provided lambda function to get guard failure reason.
+  py::function _print_failure_fn;
+};
+
+/**
+ * This is an example of relational guard and checks tensor X is not tensor Y.
+ * We construct one leaf guard and install it at as a leaf of two guard managers
+ * (one for X and another for Y). Therefore, this guard is run twice. In the
+ * first invocation, it saves the first value and returns True. In the second
+ * invocation, it compares the saved value with the new value and returns True
+ * if they do not alias.
+ */
+class NoTensorAliasingGuard : public LeafGuard {
+ public:
+  NoTensorAliasingGuard() : _is_first_call(true) {}
```

Comment:
How does this get reset back to false after the guards run?

We might need a separate GuardCheckState object (that only lives while checks are running).

Reply:
I added a virtual method - `reset_state_on_guard_failure` on the relational guards. This is the flow

1) The GuardManager has one more private member now - `_relational_guard_resetters`, a vector or Relational Guard pointers.
2) While installing a no_tensor_aliasing guard - we find the lowest common ancestor (lca) of the two guard managers. The lca node has the responsibility to call the `reset_state_on_guard_failure` of the newly added guard. This is done by registering the `_relational_guard_resetters` on the lca node.
3) On guard failure, GuardManager iterates through the resetters and call the above method.

You can review beginning from `install_no_tensor_aliasing_guard` method.

Reply:
Discussed offline. Resolving this for now.